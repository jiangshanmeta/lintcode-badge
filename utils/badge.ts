export const contentPresets = {
    rank: {
        presetLabel: "Rank",
        badgeLabel: "Rank",
        badgeValue: "rank",
        badgeValueDisplay: "Rank",
    },
    solved: {
        presetLabel: "Solved",
        badgeLabel: "Solved",
        badgeValue: "solved",
        badgeValueDisplay: "Solved",
    },
    "solved-over-total": {
        presetLabel: "Solved over total",
        badgeLabel: "Solved",
        badgeValue: "solvedOverTotal",
        badgeValueDisplay: "Solved over total",
    },
    "solved-percentage": {
        presetLabel: "Solved percentage",
        badgeLabel: "Solved",
        badgeValue: "solvedPercentage",
        badgeValueDisplay: "Solved percentage",
    },
} as const;

export const styles = {
    plastic: { width: "99px", height: "18px" },
    flat: { width: "81px", height: "20px" },
    "flat-square": { width: "125px", height: "20px" },
    "for-the-badge": { width: "210px", height: "28px" },
    social: { width: "98px", height: "20px" },
} as const;

export type ContentPresetName = keyof typeof contentPresets;

type Style = keyof typeof styles;

export interface Badge {
  username: string;
  style: Style;
  labelColor: string;
  color: string;
  contentPreset: ContentPresetName | "custom";
  label: string;
  value: ContentPresetName;
  showLogo: boolean;
  logoColor: string;
}

export const DEFAULT_BADGE: Badge = {
    username: "jiangshanmeta",
    style: "for-the-badge",
    labelColor: "black",
    color: "#12B4FF",
    contentPreset: "solved-over-total",
    label: "Solved",
    value: "solved-over-total",
    showLogo: false,
    logoColor: "yellow",
};


export const getUrl = (badge: Badge): string =>
    `https://img.shields.io/badge/dynamic/json?style=${
        badge.style
    }&labelColor=${encodeURIComponent(
        badge.labelColor
    )}&color=${encodeURIComponent(badge.color)}&label=${encodeURIComponent(
        badge.label
    )}&query=${
        contentPresets[badge.value].badgeValue
    }&url=https%3A%2F%2Flintcode-badge.vercel.app%2Fapi%2Fusers%2F${encodeURIComponent(
        badge.username
    )}`;

export const getMarkdown = (badge: Badge): string =>
    `[![LeetCode user ${badge.username}](${getUrl(
        badge
    )})](https://lintcode.com/user/${badge.username}/)`;
