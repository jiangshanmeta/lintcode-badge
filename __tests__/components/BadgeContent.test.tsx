import {
    fireEvent,
    render, 
} from '@testing-library/react';

import BadgeContent from 'components/BadgeContent';

import { DEFAULT_BADGE } from 'utils/badge';


describe('BadgeContent Component',()=>{
    beforeEach(()=>{
        jest.resetModules()
    })

    it('snapshot',()=>{
        const setBadge = jest.fn();

        const {
            container,
            rerender
        } = render(<BadgeContent badge={DEFAULT_BADGE} setBadge={setBadge} />)

        expect(container.innerHTML).toMatchSnapshot()

        rerender(<BadgeContent badge={{
            ...DEFAULT_BADGE,
            contentPreset:'custom'
        }} setBadge={setBadge}  />)

        expect(container.innerHTML).toMatchSnapshot()
    })

    describe('should handle preset change',()=>{


        it('when target is not custom',()=>{

            jest.doMock('@material-ui/core',()=>{
                return {
                    __esModule: true,
                    FormControl:(props)=>{
                        return props.children;
                    },
                    FormLabel:()=>{
                        return <div></div>
                    },
                    InputLabel:()=>{
                        return <div></div>
                    },
                    MenuItem:()=>{
                        return <div></div>
                    },
                    Select:(props)=>{
                        return <div id={props.id} onClick={()=>props.onChange({target:{value:'solved-percentage'}})  }></div>
                    },
                    TextField:()=>{
                        return <div></div>
                    },
    
                }
            
            
            })
    
            return import('@material-ui/core').then(()=>{
    
                const {default:Component} = require('components/BadgeContent')
                const setBadge = jest.fn();
    
                const {
                    container,
                } = render(<Component badge={DEFAULT_BADGE} setBadge={setBadge} />)

                fireEvent.click(container.querySelector('#badge-content-preset-select'))
    
                expect(setBadge).toBeCalledTimes(1)
    
                const firstCall = setBadge.mock.calls[0][0](DEFAULT_BADGE)
            
                expect(firstCall).toEqual({
                    ...DEFAULT_BADGE,
                    contentPreset:'solved-percentage',
                    label:'Solved',
                    value:'solved-percentage'
                })
                
            })
            
        })

        it('when target is custom',()=>{
            jest.doMock('@material-ui/core',()=>{
                return {
                    __esModule: true,
                    FormControl:(props)=>{
                        return props.children;
                    },
                    FormLabel:()=>{
                        return <div></div>
                    },
                    InputLabel:()=>{
                        return <div></div>
                    },
                    MenuItem:()=>{
                        return <div></div>
                    },
                    Select:(props)=>{
                        return <div id={props.id} onClick={()=>props.onChange({target:{value:'custom'}})  }></div>
                    },
                    TextField:()=>{
                        return <div></div>
                    },
    
                }
            
            
            })
    
            return import('@material-ui/core').then(()=>{
    
                const {default:Component} = require('components/BadgeContent')
                const setBadge = jest.fn();
    
                const {
                    container,
                } = render(<Component badge={DEFAULT_BADGE} setBadge={setBadge} />)
    
                fireEvent.click(container.querySelector('#badge-content-preset-select'))
    
                expect(setBadge).toBeCalledTimes(1)
    
                const firstCall = setBadge.mock.calls[0][0](DEFAULT_BADGE)
            
                expect(firstCall).toEqual({
                    ...DEFAULT_BADGE,
                    contentPreset:'custom',
                })
                
            })

        })

    })
 

    it('Displayed value change works',()=>{
        jest.doMock('@material-ui/core',()=>{
            return {
                __esModule: true,
                FormControl:(props)=>{
                    return props.children;
                },
                FormLabel:()=>{
                    return <div></div>
                },
                InputLabel:()=>{
                    return <div></div>
                },
                MenuItem:()=>{
                    return <div></div>
                },
                Select:(props)=>{
                    return <div id={props.id} onClick={()=>props.onChange({target:{value:'solvedPercentage'}})  }></div>
                },
                TextField:()=>{
                    return <div></div>
                },

            }
        
        
        })

        return import('@material-ui/core').then(()=>{

            const {default:Component} = require('components/BadgeContent')
            const setBadge = jest.fn();

            const initialBadge = {
                ...DEFAULT_BADGE,
                contentPreset:'custom'
            }

            const {
                container,
            } = render(<Component badge={initialBadge} setBadge={setBadge} />)

            fireEvent.click(container.querySelector('#badge-value-select'))

            expect(setBadge).toBeCalledTimes(1)

            const firstCall = setBadge.mock.calls[0][0](initialBadge)
        
            expect(firstCall).toEqual({
                ...DEFAULT_BADGE,
                contentPreset:'custom',
                value:'solvedPercentage'
            })
            
        })

    })

    it('Label change works',()=>{
        jest.doMock('@material-ui/core',()=>{
            return {
                __esModule: true,
                FormControl:(props)=>{
                    return props.children;
                },
                FormLabel:()=>{
                    return <div></div>
                },
                InputLabel:()=>{
                    return <div></div>
                },
                MenuItem:()=>{
                    return <div></div>
                },
                Select:(props)=>{
                    return <div></div>
                },
                TextField:(props)=>{
                    return <div id={props.id} onClick={()=>props.onChange({target:{value:'sakura'}}) }></div>
                },
            }
        
        
        })

        return import('@material-ui/core').then(()=>{

            const {default:Component} = require('components/BadgeContent')
            const setBadge = jest.fn();

            const {
                container,
            } = render(<Component badge={DEFAULT_BADGE} setBadge={setBadge} />)

            fireEvent.click(container.querySelector('#badge-label'))

            expect(setBadge).toBeCalledTimes(1)

            const firstCall = setBadge.mock.calls[0][0](DEFAULT_BADGE)
        
            expect(firstCall).toEqual({
                ...DEFAULT_BADGE,
                label:'sakura'
            })
            
        })

    })


})