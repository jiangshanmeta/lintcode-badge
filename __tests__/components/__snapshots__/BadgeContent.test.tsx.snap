// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadgeContent Component snapshot 1`] = `"<fieldset class="MuiFormControl-root"><legend class="MuiFormLabel-root">Content</legend><div class="MuiFormControl-root" style="margin-top: 16px;"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled" data-shrink="true" id="badge-content-preset-label">Preset</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"><div class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input" tabindex="0" role="button" aria-haspopup="listbox" aria-labelledby="badge-content-preset-label badge-content-preset-select" id="badge-content-preset-select">Solved over total</div><input aria-hidden="true" tabindex="-1" class="MuiSelect-nativeInput" value="solved-over-total"><svg class="MuiSvgIcon-root MuiSelect-icon" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M7 10l5 5 5-5z"></path></svg></div></div><div class="MuiFormControl-root MuiTextField-root" style="margin-top: 24px;"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-disabled Mui-disabled MuiFormLabel-filled" data-shrink="true" for="badge-label" id="badge-label-label">Label</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-formControl MuiInput-formControl"><input aria-invalid="false" disabled="" id="badge-label" type="text" class="MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled" value="Solved"></div></div><div class="MuiFormControl-root" style="margin-top: 24px;"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled" data-shrink="true" id="badge-value-label">Displayed value</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-formControl MuiInput-formControl"><div class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled" role="button" aria-disabled="true" aria-haspopup="listbox" aria-labelledby="badge-value-label badge-value-select" id="badge-value-select">Solved over total</div><input aria-hidden="true" tabindex="-1" class="MuiSelect-nativeInput" value="solved-over-total"><svg class="MuiSvgIcon-root MuiSelect-icon Mui-disabled" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M7 10l5 5 5-5z"></path></svg></div></div></fieldset>"`;

exports[`BadgeContent Component snapshot 2`] = `"<fieldset class="MuiFormControl-root"><legend class="MuiFormLabel-root">Content</legend><div class="MuiFormControl-root" style="margin-top: 16px;"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled" data-shrink="true" id="badge-content-preset-label">Preset</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"><div class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input" tabindex="0" role="button" aria-haspopup="listbox" aria-labelledby="badge-content-preset-label badge-content-preset-select" id="badge-content-preset-select">Solved over total</div><input aria-hidden="true" tabindex="-1" class="MuiSelect-nativeInput" value="solved-over-total"><svg class="MuiSvgIcon-root MuiSelect-icon" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M7 10l5 5 5-5z"></path></svg></div></div><div class="MuiFormControl-root MuiTextField-root" style="margin-top: 24px;"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled" data-shrink="true" for="badge-label" id="badge-label-label">Label</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"><input aria-invalid="false" id="badge-label" type="text" class="MuiInputBase-input MuiInput-input" value="Solved"></div></div><div class="MuiFormControl-root" style="margin-top: 24px;"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled" data-shrink="true" id="badge-value-label">Displayed value</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"><div class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input" role="button" aria-haspopup="listbox" aria-labelledby="badge-value-label badge-value-select" id="badge-value-select" tabindex="0">Solved over total</div><input aria-hidden="true" tabindex="-1" class="MuiSelect-nativeInput" value="solved-over-total"><svg class="MuiSvgIcon-root MuiSelect-icon" focusable="false" viewBox="0 0 24 24" aria-hidden="true"><path d="M7 10l5 5 5-5z"></path></svg></div></div></fieldset>"`;
